In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

%matplotlib inline

In [2]:
import pickle 

# Define the filename of the pickled file filename = 'data.pickle' 
# Unpickle the object from the file with open(filename, 'rb') as f: unpickled_object = pickle.load(f) 
# Print the unpickled object print(unpickled_object)

filename = 'control_confirmed.pkl'
with open(filename, 'rb') as f: 
    control_confirmed = pickle.load(f) 
control_confirmed

0  1
0      4033851  3
1      6651403  1
2      2105948  1
3      9814849  1
4      4446011  1
...        ... ..
15429  1011862  4
15430  8930684  1
15431  4840338  3
15432  4210290  1
15433  1684561  1

[15434 rows x 2 columns]

In [3]:
control_confirmed.rename(columns = {0:'client_id', 1:'count_of_steps'}, inplace=True)
control_confirmed

client_id  count_of_steps
0       4033851               3
1       6651403               1
2       2105948               1
3       9814849               1
4       4446011               1
...         ...             ...
15429   1011862               4
15430   8930684               1
15431   4840338               3
15432   4210290               1
15433   1684561               1

[15434 rows x 2 columns]

In [4]:
filename = 'how_long_in_step.pkl'
with open(filename, 'rb') as f: 
    time_on_steps = pickle.load(f) 
time_on_steps

0        1  2                   3  \
0        68804396_74737424429_397490  1000062  0 2017-04-06 15:38:49   
1        68804396_74737424429_397490  1000062  1 2017-04-06 15:41:51   
2        68804396_74737424429_397490  1000062  2 2017-04-06 15:42:01   
3        68804396_74737424429_397490  1000062  3 2017-04-06 15:42:50   
4        68804396_74737424429_397490  1000062  4 2017-04-06 15:54:49   
...                              ...      ... ..                 ...   
320220  870243567_56915814033_814203  9999729  4 2017-05-08 16:09:40   
320221   99583652_41711450505_426179  9999729  0 2017-04-05 13:40:49   
320222   99583652_41711450505_426179  9999729  1 2017-04-05 13:41:04   
320223  472154369_16714624241_585315  9999832  0 2017-05-16 16:46:03   
320224  472154369_16714624241_585315  9999832  1 2017-05-16 16:46:11   

                         4      5        6  
0      2017-04-06 15:41:51  182.0  Control  
1      2017-04-06 15:42:01   10.0  Control  
2      2017-04-06 15:42:50   49.0  Control  
3      2017-04-06 15:54:49  719.0  Control  
4                      NaT    NaN  Control  
...                    ...    ...      ...  
320220                 NaT    NaN     Test  
320221 2017-04-05 13:41:04   15.0     Test  
320222                 NaT    NaN     Test  
320223 2017-05-16 16:46:11    8.0     Test  
320224                 NaT    NaN     Test  

[320225 rows x 7 columns]

In [5]:
#rename columns to be descriptive

time_on_steps.rename(columns = {0:'visit_id',1:'client_id',2:'process_step',3:'date_time',4:'next_step_time', 5:'time_spent_sec', 6:'Variation'},inplace=True)
time_on_steps

visit_id client_id process_step  \
0        68804396_74737424429_397490   1000062            0   
1        68804396_74737424429_397490   1000062            1   
2        68804396_74737424429_397490   1000062            2   
3        68804396_74737424429_397490   1000062            3   
4        68804396_74737424429_397490   1000062            4   
...                              ...       ...          ...   
320220  870243567_56915814033_814203   9999729            4   
320221   99583652_41711450505_426179   9999729            0   
320222   99583652_41711450505_426179   9999729            1   
320223  472154369_16714624241_585315   9999832            0   
320224  472154369_16714624241_585315   9999832            1   

                 date_time      next_step_time  time_spent_sec Variation  
0      2017-04-06 15:38:49 2017-04-06 15:41:51           182.0   Control  
1      2017-04-06 15:41:51 2017-04-06 15:42:01            10.0   Control  
2      2017-04-06 15:42:01 2017-04-06 15:42:50            49.0   Control  
3      2017-04-06 15:42:50 2017-04-06 15:54:49           719.0   Control  
4      2017-04-06 15:54:49                 NaT             NaN   Control  
...                    ...                 ...             ...       ...  
320220 2017-05-08 16:09:40                 NaT             NaN      Test  
320221 2017-04-05 13:40:49 2017-04-05 13:41:04            15.0      Test  
320222 2017-04-05 13:41:04                 NaT             NaN      Test  
320223 2017-05-16 16:46:03 2017-05-16 16:46:11             8.0      Test  
320224 2017-05-16 16:46:11                 NaT             NaN      Test  

[320225 rows x 7 columns]

In [6]:
# Step 1: Filter DataFrame for 'start' steps
start_step = time_on_steps[time_on_steps['process_step'] == '0']

# Step 2: Group by 'client_id'
grouped = start_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_start = grouped['time_spent_sec'].mean()

average_time_start = pd.DataFrame(average_time_spent_start).reset_index()
average_time_start.describe()

time_spent_sec
count    46367.000000
mean        47.381256
std        107.659987
min          0.000000
25%          9.000000
50%         18.181818
75%         45.000000
max       5610.500000

In [7]:
# Step 1: Filter DataFrame for 'step 1' steps
step_1 = time_on_steps[time_on_steps['process_step'] == '1']

# Step 2: Group by 'client_id'
grouped = step_1.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_1 = grouped['time_spent_sec'].mean()

average_time_step_1 = pd.DataFrame(average_time_spent_1).reset_index()
average_time_step_1.describe()

time_spent_sec
count     42310.00000
mean         48.64149
std          86.14811
min           1.00000
25%          14.00000
50%          27.00000
75%          53.00000
max        3800.00000

In [8]:
# Step 1: Filter DataFrame for 'step 2' steps
step_2 = time_on_steps[time_on_steps['process_step'] == '2']

# Step 2: Group by 'client_id'
grouped = step_2.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_2 = grouped['time_spent_sec'].mean()

average_time_step_2 = pd.DataFrame(average_time_spent_2).reset_index()
average_time_step_2.describe()

time_spent_sec
count    39560.000000
mean        94.098579
std        144.962566
min          0.000000
25%         43.000000
50%         69.000000
75%        111.000000
max      21763.000000

In [9]:
# Step 1: Filter DataFrame for 'step 3' steps
step_3 = time_on_steps[time_on_steps['process_step'] == '3']

# Step 2: Group by 'client_id'
grouped = step_3.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_3 = grouped['time_spent_sec'].mean()

average_time_step_3 = pd.DataFrame(average_time_spent_3).reset_index()
average_time_step_3.describe()

time_spent_sec
count    36267.000000
mean       124.708733
std        177.578709
min          3.000000
25%         37.000000
50%         67.000000
75%        135.000000
max       6164.000000

In [10]:
time_spent_on_confirm = time_on_steps[time_on_steps['process_step'] == '4'][['client_id','process_step','time_spent_sec']]
time_spent_on_confirm

client_id process_step  time_spent_sec
4        1000062            4             NaN
11       1000353            4             NaN
18       1000536            4             NaN
23       1000672            4             NaN
28       1001101            4             NaN
...          ...          ...             ...
320190   9998851            4             NaN
320196   9998964            4             NaN
320201   9999009            4             NaN
320209   9999400            4             NaN
320220   9999729            4             NaN

[43082 rows x 3 columns]

In [11]:
time_spent_on_confirm['client_id'].duplicated().sum()

8961

In [12]:
# Step 1: Filter DataFrame for 'confirm' steps
confirm_step = time_on_steps[time_on_steps['process_step'] == '4']

# Step 2: Group by 'client_id'
grouped = confirm_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_confirm = grouped['time_spent_sec'].mean()

average_time_step_confirm = pd.DataFrame(average_time_spent_confirm).reset_index()
average_time_step_confirm.describe()

time_spent_sec
count     4196.000000
mean       194.523520
std        364.111482
min          0.000000
25%         38.000000
50%         84.000000
75%        193.250000
max       9396.000000

## CONCLUSION

> Started 50,076

> Confirmed 34,121

> 15,955 clients abandoned the process

> (15955/50076)*100 = 31.9% (abandonment rate)

In [13]:
average_time_step_confirm.isnull().sum()

client_id             0
time_spent_sec    29925
dtype: int64

In [14]:
#No time spent on 4 because that's when the completed the process... but there are 4,196 rows that do have time spent on the confirmed step.

In [15]:
#Average time spent per step in general (both control and test)

In [16]:
#Start average of 47 seconds
#Step 1 average of 49 seconds
#Step 2 average of 94 seconds
#Step 3 average of 125 seconds
#Confirmed step average of 195 seconds (don't understand why)

## ______________________________________________________________________________

In [17]:
#AVERAGE TIME PER STEP PER VARIATION (Control vs Test)

In [18]:
# Step 1: Filter DataFrame for 'start' steps
start_step_control = time_on_steps[(time_on_steps['process_step'] == '0') & (time_on_steps['Variation'] == 'Control')]

# Step 2: Group by 'client_id'
grouped = start_step_control.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_start = grouped['time_spent_sec'].mean()

average_time_start_control = pd.DataFrame(average_time_spent_start).reset_index()
average_time_start_control.describe()

time_spent_sec
count    21136.000000
mean        50.138087
std        112.797421
min          0.000000
25%         11.000000
50%         23.000000
75%         47.000000
max       5610.500000

In [19]:
# Step 1: Filter DataFrame for 'start' steps
start_step_control = time_on_steps[(time_on_steps['process_step'] == '0') & (time_on_steps['Variation'] == 'Test')]

# Step 2: Group by 'client_id'
grouped = start_step_control.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_start = grouped['time_spent_sec'].mean()

average_time_start_control = pd.DataFrame(average_time_spent_start).reset_index()
average_time_start_control.describe()

time_spent_sec
count    25231.000000
mean        45.071859
std        103.105153
min          0.000000
25%          8.000000
50%         15.000000
75%         42.500000
max       3953.000000

In [20]:
# Step 1: Filter DataFrame for 'step 1' steps
step_1 = time_on_steps[(time_on_steps['process_step'] == '1') & (time_on_steps['Variation'] == 'Control')]

# Step 2: Group by 'client_id'
grouped = step_1.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_1 = grouped['time_spent_sec'].mean()

average_time_step_1 = pd.DataFrame(average_time_spent_1).reset_index()
average_time_step_1.describe()

time_spent_sec
count    19142.000000
mean        44.705405
std         91.206478
min          1.000000
25%         10.000000
50%         23.000000
75%         49.000000
max       3800.000000

In [21]:
# Step 1: Filter DataFrame for 'step 1' steps
step_1 = time_on_steps[(time_on_steps['process_step'] == '1') & (time_on_steps['Variation'] == 'Test')]

# Step 2: Group by 'client_id'
grouped = step_1.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_1 = grouped['time_spent_sec'].mean()

average_time_step_1 = pd.DataFrame(average_time_spent_1).reset_index()
average_time_step_1.describe()

time_spent_sec
count    23168.000000
mean        51.893585
std         81.591781
min          2.000000
25%         17.000000
50%         30.000000
75%         56.000000
max       2321.750000

In [22]:
# Step 1: Filter DataFrame for 'step 2' steps
step_2 = time_on_steps[(time_on_steps['process_step'] == '2') & (time_on_steps['Variation'] == 'Control')]

# Step 2: Group by 'client_id'
grouped = step_2.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_2 = grouped['time_spent_sec'].mean()

average_time_step_2 = pd.DataFrame(average_time_spent_2).reset_index()
average_time_step_2.describe()

time_spent_sec
count    17861.000000
mean        97.316052
std        189.145344
min          0.000000
25%         44.000000
50%         71.000000
75%        115.000000
max      21763.000000

In [23]:
# Step 1: Filter DataFrame for 'step 2' steps
step_2 = time_on_steps[(time_on_steps['process_step'] == '2') & (time_on_steps['Variation'] == 'Test')]

# Step 2: Group by 'client_id'
grouped = step_2.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_2 = grouped['time_spent_sec'].mean()

average_time_step_2 = pd.DataFrame(average_time_spent_2).reset_index()
average_time_step_2.describe()

time_spent_sec
count    21699.000000
mean        91.450195
std         94.068443
min          0.000000
25%         43.000000
50%         67.500000
75%        107.000000
max       1995.888889

In [24]:
# Step 1: Filter DataFrame for 'step 3' steps
step_3 = time_on_steps[(time_on_steps['process_step'] == '3') & (time_on_steps['Variation'] == 'Control')]

# Step 2: Group by 'client_id'
grouped = step_3.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_3 = grouped['time_spent_sec'].mean()

average_time_step_3 = pd.DataFrame(average_time_spent_3).reset_index()
average_time_step_3.describe()

time_spent_sec
count    16281.000000
mean       134.841019
std        177.453948
min          4.000000
25%         45.000000
50%         78.000000
75%        151.500000
max       6164.000000

In [25]:
# Step 1: Filter DataFrame for 'step 3' steps
step_3 = time_on_steps[(time_on_steps['process_step'] == '3') & (time_on_steps['Variation'] == 'Test')]

# Step 2: Group by 'client_id'
grouped = step_3.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_3 = grouped['time_spent_sec'].mean()

average_time_step_3 = pd.DataFrame(average_time_spent_3).reset_index()
average_time_step_3.describe()

time_spent_sec
count    19986.000000
mean       116.454768
std        177.257134
min          3.000000
25%         32.000000
50%         58.000000
75%        120.000000
max       3400.000000

In [26]:
# Step 1: Filter DataFrame for 'confirm' steps
confirm_step = time_on_steps[(time_on_steps['process_step'] == '4') & (time_on_steps['Variation'] == 'Control')]

# Step 2: Group by 'client_id'
grouped = confirm_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_confirm = grouped['time_spent_sec'].mean()

average_time_step_confirm = pd.DataFrame(average_time_spent_confirm).reset_index()
average_time_step_confirm.describe()

time_spent_sec
count     1408.000000
mean       165.339824
std        406.082176
min          0.000000
25%         21.187500
50%         56.250000
75%        141.250000
max       9396.000000

In [27]:
# Step 1: Filter DataFrame for 'confirm' steps
confirm_step = time_on_steps[(time_on_steps['process_step'] == '4') & (time_on_steps['Variation'] == 'Test')]

# Step 2: Group by 'client_id'
grouped = confirm_step.groupby('client_id')

# Step 3: Calculate the average time spent for each client
average_time_spent_confirm = grouped['time_spent_sec'].mean()

average_time_step_confirm = pd.DataFrame(average_time_spent_confirm).reset_index()
average_time_step_confirm.describe()

time_spent_sec
count     2788.000000
mean       209.261915
std        340.084996
min          0.000000
25%         48.000000
50%         99.687500
75%        219.166667
max       5614.000000

### CONCLUSIONS 

### The average time spent per step (including both variations):

> Start average of 47 seconds

> Step 1 average of 49 seconds

> Step 2 average of 94 seconds

> Step 3 average of 125 seconds

> Confirmed step average of 195 seconds (don't understand why)

### The average time spent per step in the CONTROL GROUP:

> Start average of 50.14 seconds

> Step 1 average of 44.71 seconds

> Step 2 average of 97.32 seconds

> Step 3 average of 134.84 seconds

> Confirmed step average of 165.34 seconds

### The average time spent per step in the TEST GROUP:

> Start average of 45.07 seconds

> Step 1 average of 51.89 seconds (LONGER THAN THE CONTROL)

> Step 2 average of 91.45 seconds

> Step 3 average of 116.45 seconds

> Confirmed step average of 209.26 seconds (LONGER THAN THE CONTROL)

### Time Spent Rate

> START : 10.8% faster in the test group

> Step 1 : 14.7% faster in the control group

> Step 2 : 6.2% faster in the test group

> Step 3 : 14.7% faster in the test group

> Confirmed : 22.5% faster in the control group (was there more information once completed?)

## __________________________________________________________________________________

In [28]:
time_on_steps.to_csv('control_confirmed.csv', index=False)